designs/mixins.md

```diff
@@ -573,34 +510,33 @@ structure A2 {
     A: String
 }
 
-structure Invalid with A1, A2 {}
+structure Invalid + A1 + A2 {}
 ```
 
 
 ### Mixins in the IDL
 
 To support mixins, `structure_statement` and `union_statement` ABNF rules
-will be updated to contain an optional `with` clause that comes after the
-shape name and before `{`. If present, the `with` clause MUST contain one or
-more shape IDs, and each shape MUST target a valid shape marked with the
-`@mixin` trait. The `(inherit)` keyword also requires updating the Smithy
-IDL grammar for structure and union shapes.
+will be updated to contain an optional series of `add_mixin` productions
+that comes after the shape name and before `{`. Each `+` in the
+`add_mixin` MUST be followed by a single shape ID, and each shape MUST
+target a valid shape marked with the `@mixin` trait. Any number of mixins can
+be added to a shape.
 
 ```
-structure_statement = "structure" ws `identifier` ws [with_clause] structure_members
-structure_members: "{" `ws` *(`structure_members_kvp` `ws`) "}"
-structure_members_kvp: `trait_statements` `identifier` `ws`
-    ":" `ws` `structure_and_union_member_target` [`required_member_sugar`]
+structure_statement = "structure" ws `identifier` ws *add_mixin structure_members
+union_statement = "union" ws `identifier` ws *add_mixin union_members
+add_mixin = "+" `ws` `shape_id`
+```
 
-with_clause = "with" ws 1*shape_id ws
-structure_and_union_member_target: `shape_id` / `inherit_keyword`
-inherit_keyword: "(inherit)"
+#### Why `+` and not `with`?
+
+By using a symbol like `+`, there is more flexibility to evolve the IDL in
+the future while avoiding potential grammar ambiguities. Given that `+`
+conveys that mixins are added to shapes, the improved future grammar
+flexibility gives `+` a slight edge over other terms like `with` that are
+valid `identifier` productions.
```

Comment:
I personally like `+`, and I don't see `with` as more clear than `+`. To me, we are defining a shape that adds another shape into its definition, so `+` made a ton of sense, and it doesn't imply some kind of synthetic shape.

> I don't think the fact that with is a valid identifier is important. So is structure and all of our other keywords.

I sat down an started writing out the grammar, and I realized that if we just rewrite the `operation_statement` grammar, then we can remove the ambiguity from likely any inline input/output option. In the way we _currently_ model operations, it would be ambiguous (we just use `object_node` and then refer to the members defined elsewhere in the spec). But if we explicitly define the grammar for operation key value pairs (we should), then the ambiguity could go away, because "with" is not one of the key-value pairs of an operation.

What if we did something like this instead? Do you like this more? I think I do.

```
structure SomeStructure {
    + SomeMixin1
    + SomeMixin2

    foo: String
}
```

Reply:
Yeah I definitely like that better. My main problem with the `structure A + B {}` is the lack of distinction between `A` and `B`. `with` is an explicitly directional statement, so it's very clear in `structure A with B {}` that `B` is being added to `A`. `+` is commutative, so there isn't that distinction. Putting it in the body adds that distinction

Reply:
We polled our internal Smithy developers and the voting landed back on the `with` syntax. It looks more familiar, it doesn't have the lexical implications of putting the mixins inside the structure (traits are copied onto the structure, and some worried that if we put them inside then it implies only members are copied), and `with` felt more readable and less like code-golf.